gravitee-am-service/src/main/java/io/gravitee/am/service/impl/VerifyAttemptServiceImpl.java

```diff
@@ -0,0 +1,174 @@
+            return Completable.complete();
+        }
+        if (optionalVerifyAttempt.isPresent()) {
+            final VerifyAttempt verifyAttempt = optionalVerifyAttempt.get();
+            final int attempts = verifyAttempt.getAttempts() + 1;
+            if (attempts >= accountSettings.getMfaChallengeMaxAttempts()) {
+                verifyAttempt.setAllowRequest(false);
+                verifyAttempt.setAttempts(accountSettings.getMfaChallengeMaxAttempts());
+            } else {
+                verifyAttempt.setAttempts(attempts);
+                verifyAttempt.setAllowRequest(true);
+            }
+            verifyAttempt.setUpdatedAt(new Date());
+
+            return verifyAttemptRepository.update(verifyAttempt).ignoreElement();
+
+        } else {
+            final VerifyAttempt verifyAttempt = new VerifyAttempt();
+            verifyAttempt.setUserId(userId);
+            verifyAttempt.setFactorId(factorId);
+            verifyAttempt.setClient(client.getId());
+            verifyAttempt.setReferenceId(domain.getId());
+            verifyAttempt.setReferenceType(ReferenceType.DOMAIN);
+            final int attempts = 1;
+            verifyAttempt.setAttempts(attempts);
+            verifyAttempt.setAllowRequest(attempts < accountSettings.getMfaChallengeMaxAttempts());
+            verifyAttempt.setCreatedAt(new Date());
+            verifyAttempt.setUpdatedAt(verifyAttempt.getCreatedAt());
+
+            return verifyAttemptRepository.create(verifyAttempt).ignoreElement();
+        }
+    }
+
+    @Override
+    public Completable delete(String id) {
+        LOGGER.debug("delete verify attempt id: {}", id);
+        return verifyAttemptRepository.delete(id);
+    }
+
+    @Override
+    public Completable deleteByUser(User user) {
+        LOGGER.debug("deleteByUser userID: {}", user.getId());
+        return verifyAttemptRepository.deleteByUser(user.getId());
+    }
+
+    @Override
+    public Completable deleteByDomain(Domain domain, ReferenceType referenceType) {
+        LOGGER.debug("deleteByDomain domainId: {}", domain.getId());
+        return verifyAttemptRepository.deleteByDomain(domain.getId(), referenceType);
+    }
+
+    @Override
+    public boolean shouldSendEmail(Client client, Domain domain) {
+        final AccountSettings accountSettings = AccountSettings.getInstance(domain, client);
+        return accountSettings.isMfaChallengeSendVerifyAlertEmail();
+    }
+
+    private Maybe<VerifyAttempt> getVerifyAttemptIfExists(VerifyAttemptCriteria criteria, AccountSettings accountSettings) {
+        return verifyAttemptRepository.findByCriteria(criteria)
+                .map(Optional::of)
+                .defaultIfEmpty(Optional.empty())
+                .flatMap(optionalVerifyAttempt -> {
+                    if (optionalVerifyAttempt.isEmpty()) {
+                        return Maybe.empty();
+                    } else {
```

Comment:
If the `flatMap` method returns `Maybe.empty` when `optionalVerifyAttempt` is empty, I think we can directly do the flatMap after the findByCriteria.

```
verifyAttemptRepository.findByCriteria(criteria).flatMap(verifyAttempt -> {
final Date resetTime = new Date(verifyAttempt.getUpdatedAt().getTime() + (accountSettings.getMfaChallengeAttemptsResetTime() * 1000));
                        final Date currentDate = new Date();

                        if (currentDate.getTime() > resetTime.getTime()) {
                            verifyAttempt.setAttempts(0);
                            verifyAttempt.setAllowRequest(true);
                        } else {
                            verifyAttempt.setAllowRequest(verifyAttempt.getAttempts() < accountSettings.getMfaChallengeMaxAttempts());
                        }
                        return Maybe.just(verifyAttempt);
})
```

if findByCriteria return a Maybe.empty, the flatMap will not be executed.

Reply:
changed as suggested